documentation/writing_tests.md

```diff
@@ -0,0 +1,1567 @@
+
+### Karma
+
+Karma is a tool that allows us to run the same TypeScript tests we run on node, but on the browsers. The Karma runner as a tool has a considerably low barrier of entry, but it is also pretty extensive. We recommend going through the following resources to learn more about Karma:
+
+- About Karma's [configuration file](http://karma-runner.github.io/4.0/config/configuration-file.html).
+- About Karma's handling of [files](http://karma-runner.github.io/4.0/config/files.html).
+- About Karma's support for [browsers](http://karma-runner.github.io/4.0/config/browsers.html).
+- About Karma's [preprocessors](http://karma-runner.github.io/4.0/config/preprocessors.html).
+- About Karma's [plugins](http://karma-runner.github.io/4.0/config/plugins.html).
+
+#### Karma in our dependencies
+
+If you want to add `karma` as a dev dependency to a new project, please use [Rush](https://rushjs.io/) since we're most likely using the same version of Karma in more projects. First make sure you are in the root folder of that project, then you can run the following command:
+
+```
+rush add --dev -p karma
+```
+
+To fulfill our needs, we use Karma with some plugins. They're the following:
+
+- [`karma-coverage`](https://www.npmjs.com/package/karma-coverage):
+  Generate code coverage using [Istanbul](https://github.com/gotwarlost/istanbul).
+- [`karma-env-preprocessor`](https://www.npmjs.com/package/karma-env-preprocessor):
+  It's a Karma preprocessor which makes environment variables available to our tests.
+- [`karma-json-preprocessor`](https://www.npmjs.com/package/karma-json-preprocessor):
+  It's a Karma preprocessor for converting JSON files into JS variables.
+- [`karma-json-to-file-reporter`](https://www.npmjs.com/package/karma-json-to-file-reporter):
+  It's a Karma reporter that save JSON messages from log to file. We use this to create real local files while executing tests in the browser. It's specifically used by [the recorder](#the-recorder).
+- [`karma-junit-reporter`](https://www.npmjs.com/package/karma-junit-reporter):
+  It's a Karma reporter for the [JUnit XML format](https://llg.cubic.org/docs/junit/).
+- [`karma-mocha`](https://www.npmjs.com/package/karma-mocha):
+  It's an adapter for the [Mocha](http://mochajs.org/) testing framework.
+- [`karma-mocha-reporter`](https://www.npmjs.com/package/karma-mocha-reporter):
+  It's a Karma reporter plugin with mocha style logging.
+- [`karma-remap-istanbul`](https://www.npmjs.com/package/karma-remap-istanbul):
+  A Karma reporter plugin that remaps the code coverage report to point to the typescript sources using sourcemappings.
+- [`karma-chrome-launcher`](https://www.npmjs.com/package/karma-chrome-launcher):
+  Launcher for Google Chrome, Google Chrome Canary and Google Chromium.
+
+You can install them with:
+
+```
+rush add --dev -p karma
```

Comment:
Does this make sure that the version being installed is in sync with the rest of the SDKs in the repo?

Reply:
Rush tries to do this by default:

> If no version is specified, a version will be automatically detected (typically
either the latest version or a version that won't break the
"ensureConsistentVersions" policy).

From: https://rushjs.io/pages/commands/rush_add/